src/core/border_router/routing_manager.cpp

```diff
@@ -0,0 +1,700 @@
+    VerifyOrExit(Get<Mle::MleRouter>().IsAttached());
+    VerifyOrExit(IsValidOmrPrefix(aOmrPrefix));
+
+    IgnoreError(Get<NetworkData::Local>().RemoveOnMeshPrefix(aOmrPrefix));
+    Get<NetworkData::Notifier>().HandleServerDataUpdated();
+
+exit:
+    return;
+}
+
+Ip6::Prefix RoutingManager::EvaluateOnLinkPrefix(void)
+{
+    Ip6::Prefix newOnLinkPrefix;
+
+    newOnLinkPrefix.Clear();
+
+    // We don't evalute on-link prefix if we are doing
+    // Router Discovery or we have already discovered some
+    // on-link prefixes.
+    VerifyOrExit(!mRouterSolicitTimer.IsRunning());
+    if (IsValidOnLinkPrefix(mDiscoveredOnLinkPrefix))
+    {
+        otLogInfoBr("EvaluateOnLinkPrefix: there is already on-link prefix %s on interface %s",
+                    mDiscoveredOnLinkPrefix.ToString().AsCString(), mInfraIfName);
+        ExitNow();
+    }
+
+    if (IsValidOnLinkPrefix(mAdvertisedOnLinkPrefix))
+    {
+        newOnLinkPrefix = mAdvertisedOnLinkPrefix;
+    }
+    else
+    {
+        newOnLinkPrefix = mLocalOnLinkPrefix;
+    }
+
+exit:
+    return newOnLinkPrefix;
+}
+
+void RoutingManager::EvaluateRoutingPolicy(void)
+{
+    Ip6::Prefix newOnLinkPrefix;
+    Ip6::Prefix newOmrPrefix;
+
+    VerifyOrExit(IsInitialized() && Get<Mle::MleRouter>().IsAttached());
+
+    otLogInfoBr("evaluating routing policy");
+
+    newOnLinkPrefix = EvaluateOnLinkPrefix();
+    newOmrPrefix    = EvaluateOmrPrefix();
+
+    if (newOmrPrefix == mLocalOmrPrefix)
+    {
+        if (!IsValidOmrPrefix(mAdvertisedOmrPrefix))
+        {
+            otLogInfoBr("publish new OMR prefix in Thread network");
+            PublishOmrPrefix(newOmrPrefix);
+        }
+    }
+    else
+    {
+        if (IsValidOmrPrefix(mAdvertisedOmrPrefix))
+        {
+            otLogInfoBr("there is already OMR prefix in the Thread network, stop publishing");
+            UnpublishOmrPrefix(mAdvertisedOmrPrefix);
+            mAdvertisedOmrPrefix.Clear();
+        }
+    }
+
+    SendRouterAdvertisement(newOmrPrefix, newOnLinkPrefix);
+
+    mAdvertisedOnLinkPrefix = newOnLinkPrefix;
+    mAdvertisedOmrPrefix    = newOmrPrefix;
+
+exit:
+    return;
+}
+
+void RoutingManager::SendRouterSolicit(void)
+{
+    otError                         error;
+    Ip6::Address                    destAddress;
+    RouterAdv::RouterSolicitMessage routerSolicit;
+
+    VerifyOrExit(IsInitialized() && Get<Mle::MleRouter>().IsAttached());
+
+    destAddress.SetToLinkLocalAllRoutersMulticast();
+    error = otPlatInfraIfSendIcmp6(mInfraIfIndex, &destAddress, reinterpret_cast<const uint8_t *>(&routerSolicit),
+                                   sizeof(routerSolicit));
+
+    if (error == OT_ERROR_NONE)
+    {
+        mRouterSolicitTimer.Start(kRtrSolicitationInterval * 1000);
+        otLogInfoBr("sent Router Solicitation, timer scheduled in %u seconds", kRtrSolicitationInterval);
+    }
+    else
+    {
+        otLogCritBr("failed to send Router Solicitation: %s", otThreadErrorToString(error));
```

Comment:
Should we retry after a while? 
The code seems to do nothing if the failed to send the first RS message.

Reply:
SGTM.